src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/RSACng.ImportExport.cs

```diff
@@ -0,0 +1,212 @@
+            //     -- Private only --
+            //     byte[cbPrime1]       prime1              - P
+            //     byte[cbPrime2]       prime2              - Q
+            //     byte[cbPrime1]       exponent1           - DP
+            //     byte[cbPrime2]       exponent2           - DQ
+            //     byte[cbPrime1]       coefficient         - InverseQ
+            //     byte[cbModulus]      privateExponent     - D
+            //
+            byte[] tempMagic = new byte[4];
+            tempMagic[0] = rsaBlob[0]; tempMagic[1] = rsaBlob[1]; tempMagic[2] = rsaBlob[2]; tempMagic[3] = rsaBlob[3];
+            KeyBlobMagicNumber magic = (KeyBlobMagicNumber)BitConverter.ToInt32(tempMagic, 0);
+            //Check the magic value in key blob header. If blob does not have required magic 
+            // then it trhows Cryptographic exception
+            CheckMagicValueOfKey(magic, includePrivateParameters);
+
+            unsafe
+            {
+                // Fail-fast if a rogue provider gave us a blob that isn't even the size of the blob header.
+                if (rsaBlob.Length < sizeof(BCRYPT_RSAKEY_BLOB))
+                    throw ErrorCode.E_FAIL.ToCryptographicException();
+
+                fixed (byte* pRsaBlob = rsaBlob)
+                {
+                    BCRYPT_RSAKEY_BLOB* pBcryptBlob = (BCRYPT_RSAKEY_BLOB*)pRsaBlob;
+
+                    int offset = sizeof(BCRYPT_RSAKEY_BLOB);
+
+                    // Read out the exponent
+                    rsaParams.Exponent = Consume(rsaBlob, ref offset, pBcryptBlob->cbPublicExp);
+                    rsaParams.Modulus = Consume(rsaBlob, ref offset, pBcryptBlob->cbModulus);
+
+                    if (includePrivateParameters)
+                    {
+                        rsaParams.P = Consume(rsaBlob, ref offset, pBcryptBlob->cbPrime1);
+                        rsaParams.Q = Consume(rsaBlob, ref offset, pBcryptBlob->cbPrime2);
+                        rsaParams.DP = Consume(rsaBlob, ref offset, pBcryptBlob->cbPrime1);
+                        rsaParams.DQ = Consume(rsaBlob, ref offset, pBcryptBlob->cbPrime2);
+                        rsaParams.InverseQ = Consume(rsaBlob, ref offset, pBcryptBlob->cbPrime1);
+                        rsaParams.D = Consume(rsaBlob, ref offset, pBcryptBlob->cbModulus);
+                    }
+                }
+            }
+
+            return rsaParams;
+        }
+
+        /// <summary>
+        ///     This function checks the magic value in the key blob header
+        /// </summary>
+        /// <param name="includePrivateParameters">Private blob if true else public key blob</param>
+        private static void CheckMagicValueOfKey(KeyBlobMagicNumber magic, bool includePrivateParameters)
+        {
+            if (includePrivateParameters)
+            {
+                if (magic != KeyBlobMagicNumber.BCRYPT_RSAPRIVATE_MAGIC && magic != KeyBlobMagicNumber.BCRYPT_RSAFULLPRIVATE_MAGIC)
```

Comment:
It seems a little weird that we'd accept BCRYPT_RSAPRIVATE_MAGIC; because we've assumed the structure should be the size of BCRYPT_RSAFULLPRIVATE_BLOB (though we called it BCRYPT_RSAKEY_BLOB).

Since I know that's what full framework is already doing I'm not sure what the implications are; just that the code smells a little fishy.